driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -405,9 +405,52 @@ private void incrementAuthErrorMetric() {
         }
     }
 
-    private UnsupportedProtocolVersionException unsupportedProtocolVersionException(ProtocolVersion triedVersion, ProtocolVersion serverProtocolVersion) {
-        logger.debug("Got unsupported protocol version error from {} for version {} server supports version {}", address, triedVersion, serverProtocolVersion);
-        return new UnsupportedProtocolVersionException(address, triedVersion, serverProtocolVersion);
+    private static final Pattern SUPPORTED_VERSIONS_PATTERN = Pattern.compile("supported versions are \\(([^)]+?)\\)");
```

Comment:
hmm now that I think of it, maybe we could simplify the algorithm to:

1. if the server version differs from the client one, then use that version;
2. otherwise upgrade to server version + 1, if possible;
3. otherwise fail.

Step 2 sounds counterintuitive, but that could actually work for all cases. Will give some more thought.